### ISSUE TITLE
Memory leak in *_generator

### ISSUE THREAD

**Issue by Dref360** [keras-team/keras#10756]
- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/keras-team/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

Since tf 1.8, the memory leak is worse. It's only happening when calling fit/evaluate multiple time so this is not likely to happen in a real setting. But this may have played a role in https://github.com/keras-team/keras/pull/10674 .

For info, tf > 1.8 uses more memory than 1.7.

# How to reproduce:
`pip install psutil`

```python
import numpy as np
import tensorflow as tf

from keras import Sequential
from keras.layers import Dense

print("Tensorflow version", tf.__version__)

model1 = Sequential()
model1.add(Dense(10, input_shape=(1000,)))

model1.add(Dense(3, activation='relu'))
model1.compile('sgd', 'mse')


def gen():
    while True:
        yield np.zeros([10, 1000]), np.ones([10, 3])


import os
import psutil

process = psutil.Process(os.getpid())
g = gen()
while True:
    print(process.memory_info().rss / float(2 ** 20))
    model1.fit_generator(g, 100, 2, use_multiprocessing=True, verbose=0)
    model1.evaluate_generator(gen(), 100, use_multiprocessing=True, verbose=0)
```

# Tensorflow 1.7
```
('Tensorflow version', '1.7.0')
160.546875
176.2421875
177.70703125
177.8671875
177.87109375
177.87890625
177.88671875
177.890625
177.8984375
177.90234375
177.90234375
177.90234375
177.90234375
177.91015625
177.9140625
177.91796875
177.91796875
177.91796875
177.91796875
177.91796875
177.91796875
177.91796875
177.93359375
```

# Tensorflow 1.8
```
('Tensorflow version', '1.9.0')
166.3125
180.30078125
182.125
182.16015625
182.22265625
182.31640625
182.3203125
182.32421875
182.3359375
182.34375
182.34765625
182.35546875
182.35546875
182.359375
182.37109375
182.38671875
182.390625
182.40234375
182.40625
182.41015625
182.41796875
182.41796875
182.41796875
182.41796875
```

**Comment by tRosenflanz** [comment]
I can confirm this behaviour across several systems + GPU memory also doesn't get properly cleaned sometimes causing OOMs. I have noticed that the worker processes don't always terminate after fitting is done/interrupted.

**Comment by tRosenflanz** [comment]
So was I correctish about it being linked to multiprocessing: https://github.com/keras-team/keras/pull/10674 ?

I might need to disable some multiprocessing in some scripts until the issue is identified.

**Comment by Dref360** [comment]
From what I found those a 2 distinct problems. I used pytest-memory-profiler.
In #10674, the processes take a lot of memory at the start (2GB), this goes OOM and then the OS kills the processes silently. 

I profiled the script above with memory_profiler and it seems that a lot of dict have no owner. And it seems like a bunch of logs. So maybe we're not cleaning them correctly.

It is to be noted that we don't see this behaviour with keras.utils.Sequence so refactoring GeneratorEnqueuer to use a Pool may be the solution.

**Comment by soulmachine** [comment]
What's the status of this issue now? I have the same issue recent days, I'm using 
a data generator inherited from `keras.utils.Sequence` with `use_multiprocessing=True`, and the memory keeps growing until the training process dies.

**Comment by Dref360** [comment]
@soulmachine this is an issue when using generators, not Sequences.

Be sure to be on the latest version of Keras. If the issue persists, please submit a reproducible script.

**Comment by mikesj-public** [comment]
Hey guys, we have a small reporducible example here which gives us a memory leak

Keras version : 2.2.4
Tensorflow 1.13.1
Ubuntu 18.04.2
CUDA Version: 10.1

`port os
import numpy as np

os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
os.environ["CUDA_VISIBLE_DEVICES"]="3"

import tensorflow as tf
from keras import backend as K
from keras import initializers
from keras.backend.tensorflow_backend import set_session
from keras.callbacks import EarlyStopping, ModelCheckpoint
from keras.layers import Input, Dense, Activation, Permute, Multiply, TimeDistributed, Layer
from keras.layers import Reshape, Lambda, Flatten, RepeatVector, Dropout, Conv2DTranspose
from keras.layers import Bidirectional, LSTM, CuDNNLSTM, BatchNormalization
from keras.layers import Reshape, Lambda, Flatten, RepeatVector, Dropout, Conv2DTranspose, UpSampling2D
from keras.layers.convolutional import Conv2D, MaxPooling2D
from keras.models import Model
from keras.optimizers import Adam
from keras.utils import Sequence


class KerasGenerator(Sequence):
	def __len__(self):
		return 10000

	def __getitem__(self, idx):
		X = np.random.rand(32, 576, 576, 1).astype(np.float32)
		y = np.round(np.random.rand(32, 576, 576, 2)).astype(np.float32)
		return X, y


def fcn(input_size=576, input_channels=3, conv_filters=32, kernel_size=(3, 3), 
		pool_size=2, nb_classes=2, learning_rate=1e-3):
	inputs = Input(shape=(input_size, input_size, input_channels))
	inner = Conv2D(conv_filters, kernel_size, padding='same',
				   activation='relu', kernel_initializer='he_normal',
				   name='conv1')(inputs)
	inner = MaxPooling2D(pool_size=(pool_size, pool_size), name='max1')(inner)
	inner = Conv2D(conv_filters*2, kernel_size, padding='same',
				   activation='relu', kernel_initializer='he_normal',
				   name='conv2')(inner)
	inner = MaxPooling2D(pool_size=(pool_size, pool_size), name='max2')(inner)
	inner = Conv2D(conv_filters*4, kernel_size, padding='same',
				   activation='relu', kernel_initializer='he_normal',
				   name='conv3')(inner)
	inner = MaxPooling2D(pool_size=(pool_size, pool_size), name='max3')(inner)
	inner = Conv2D(conv_filters*8, kernel_size, padding='same',
				   activation='relu', kernel_initializer='he_normal',
				   name='conv4')(inner)
	inner = MaxPooling2D(pool_size=(pool_size, pool_size), name='max4')(inner)
	inner = Conv2D(conv_filters*16, kernel_size, padding='same',
				   activation='relu', kernel_initializer='he_normal',
				   name='conv5')(inner)
	inner = MaxPooling2D(pool_size=(pool_size, pool_size), name='max5')(inner)
	
	pooled_size = input_size // (pool_size ** 5)
	inner = Reshape((pooled_size**2, conv_filters*16))(inner)
	inner = Bidirectional(CuDNNLSTM(128, return_sequences=True))(inner)
	inner = Dense(nb_classes)(inner)
	x = Reshape((pooled_size, pooled_size, nb_classes))(inner)

	x = Conv2DTranspose(filters=nb_classes,
						kernel_size=(64, 64),
						strides=(32, 32),
						padding='same',
						activation='softmax',
						kernel_initializer='he_normal')(x)
	model = Model(inputs=inputs, outputs=x)
	model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
	return model


with tf.device("/gpu:0"):
	K.clear_session()
	tf.reset_default_graph()
	config = tf.ConfigProto()
	config.gpu_options.allow_growth=True
	sess = tf.Session(config=config)
	set_session(sess)

	gen = KerasGenerator()
	gen_val = KerasGenerator()
	model = fcn(input_channels=1)
	model.fit_generator(gen, use_multiprocessing=True, workers=5, epochs=20,
					   validation_data=gen_val, max_queue_size=1)
`
setting use_multiprocessing = True in final lines leads to an memory leak, False has no leak.

**Comment by atotev** [comment]
https://fantashit.com/linearly-increasing-memory-with-use-multiprocessing-and-keras-sequence/#comment-254237
